File: frontend/src/api/export.js
/**
 * Export API calls.
 */

import { API_BASE_URL, apiRequest } from './client.js';
import { getAccessToken } from './tokenStorage.js';

function authHeaders(extra = {}) {
  const headers = {
    'Content-Type': 'application/json',
    ...extra,
  };
  const token = getAccessToken();
  if (token) {
    headers.Authorization = `Bearer ${token}`;
  }
  return headers;
}

/**
 * Save a quick-scan report as Markdown in scan_results/.
 *
 * @param {string} reportId
 * @returns {Promise<{ saved: object, download_filename: string }>}
 */
export async function saveMarkdownReport(reportId) {
  return apiRequest('/export/markdown', {
    method: 'POST',
    headers: authHeaders({ Accept: 'application/json' }),
    body: JSON.stringify({ report_id: reportId }),
  });
}

/**
 * Save a quick-scan report as JSON in scan_results/.
 *
 * @param {string} reportId
 * @returns {Promise<{ saved: object, download_filename: string }>}
 */
export async function saveJsonReport(reportId) {
  return apiRequest('/export/json', {
    method: 'POST',
    headers: authHeaders({ Accept: 'application/json' }),
    body: JSON.stringify({ report_id: reportId }),
  });
}

/**
 * Save a quick-scan report as CSV in scan_results/.
 *
 * @param {string} reportId
 * @returns {Promise<{ saved: object, download_filename: string }>}
 */
export async function saveCsvReport(reportId) {
  return apiRequest('/export/csv', {
    method: 'POST',
    headers: authHeaders({ Accept: 'application/json' }),
    body: JSON.stringify({ report_id: reportId }),
  });
}

/**
 * Save a quick-scan report as PDF in scan_results/.
 *
 * @param {string} reportId
 * @returns {Promise<{ saved: object, download_filename: string }>}
 */
export async function savePdfReport(reportId) {
  return apiRequest('/export/pdf', {
    method: 'POST',
    headers: authHeaders({ Accept: 'application/json' }),
    body: JSON.stringify({ report_id: reportId }),
  });
}

/**
 * List saved scan result files.
 *
 * @returns {Promise<{ items: object[], total: number }>}
 */
export function listScanResults() {
  return apiRequest('/export/scan-results');
}

/**
 * Build an authenticated URL for viewing a saved scan result.
 *
 * @param {string} filename
 * @returns {string}
 */
export function getScanResultViewPath(filename) {
  return `/scan-results/view/${encodeURIComponent(filename)}`;
}

/**
 * Download a saved scan result file to the user's machine.
 *
 * @param {string} filename
 * @returns {Promise<void>}
 */
export async function downloadScanResult(filename) {
  const response = await fetch(
    `${API_BASE_URL}/export/scan-results/${encodeURIComponent(filename)}/download`,
    {
      headers: authHeaders({ Accept: '*/*' }),
    },
  );
  if (!response.ok) {
    const body = await response.json().catch(() => ({}));
    throw new Error(body.message ?? body.detail ?? 'Download failed');
  }
  const blob = await response.blob();
  const url = URL.createObjectURL(blob);
  const link = document.createElement('a');
  link.href = url;
  link.download = filename;
  link.click();
  URL.revokeObjectURL(url);
}

/**
 * Fetch a saved scan result for inline viewing.
 *
 * @param {string} filename
 * @returns {Promise<{ contentType: string, blob: Blob, text?: string }>}
 */
export async function fetchScanResult(filename) {
  const response = await fetch(
    `${API_BASE_URL}/export/scan-results/${encodeURIComponent(filename)}`,
    {
      headers: authHeaders({ Accept: '*/*' }),
    },
  );
  if (!response.ok) {
    const body = await response.json().catch(() => ({}));
    throw new Error(body.message ?? body.detail ?? 'Failed to load scan result');
  }
  const contentType = response.headers.get('content-type') ?? '';
  if (contentType.includes('text/markdown') || filename.toLowerCase().endsWith('.md')) {
    const text = await response.text();
    return { contentType, blob: new Blob([text], { type: contentType }), text };
  }
  const blob = await response.blob();
  return { contentType, blob };
}

File: backend/app/routers/export.py
"""Export API routes."""

from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, status
from fastapi.responses import Response

from app.auth.dependencies import require_viewer
from app.models.user import User
from app.schemas.export import (
    ExportReportRequest,
    ExportSavedResponse,
    ScanResultListResponse,
)
from app.services.quick_scan_service import QuickScanError, quick_scan_service
from app.services.report_service import report_service
from app.services.scan_result_storage import ScanResultStorageError, scan_result_storage

router = APIRouter(prefix="/export", tags=["export"])


@router.get(
    "/scan-results",
    response_model=ScanResultListResponse,
    summary="List saved scan result files",
)
def list_scan_results(
    _: Annotated[User, Depends(require_viewer)],
) -> ScanResultListResponse:
    """List Markdown and PDF reports saved in scan_results/."""
    items = scan_result_storage.list_results()
    return ScanResultListResponse(items=items, total=len(items))


@router.get(
    "/scan-results/{filename}/download",
    summary="Download a saved scan result file",
)
def download_scan_result(
    filename: str,
    _: Annotated[User, Depends(require_viewer)],
) -> Response:
    """Serve a saved report as a file download."""
    try:
        if filename.lower().endswith(".pdf"):
            pdf_content = scan_result_storage.read_bytes(filename)
            return Response(
                content=pdf_content,
                media_type="application/pdf",
                headers={"Content-Disposition": f'attachment; filename="{filename}"'},
            )
        if filename.lower().endswith(".md"):
            md_content = scan_result_storage.read_text(filename)
            return Response(
                content=md_content,
                media_type="text/markdown; charset=utf-8",
                headers={"Content-Disposition": f'attachment; filename="{filename}"'},
            )
        if filename.lower().endswith(".json"):
            json_content = scan_result_storage.read_text(filename)
            return Response(
                content=json_content,
                media_type="application/json",
                headers={"Content-Disposition": f'attachment; filename="{filename}"'},
            )
        if filename.lower().endswith(".csv"):
            csv_content = scan_result_storage.read_text(filename)
            return Response(
                content=csv_content,
                media_type="text/csv; charset=utf-8",
                headers={"Content-Disposition": f'attachment; filename="{filename}"'},
            )
    except ScanResultStorageError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=str(exc),
        ) from exc
    raise HTTPException(
        status_code=status.HTTP_400_BAD_REQUEST,
        detail="Unsupported file type",
    )


@router.get(
    "/scan-results/{filename}",
    summary="View a saved scan result in the browser",
)
def view_scan_result(
    filename: str,
    _: Annotated[User, Depends(require_viewer)],
) -> Response:
    """Serve a saved report for inline browser viewing."""
    try:
        if filename.lower().endswith(".pdf"):
            pdf_content = scan_result_storage.read_bytes(filename)
            return Response(
                content=pdf_content,
                media_type="application/pdf",
                headers={"Content-Disposition": f'inline; filename="{filename}"'},
            )
        if filename.lower().endswith(".md"):
            md_content = scan_result_storage.read_text(filename)
            return Response(
                content=md_content,
                media_type="text/markdown; charset=utf-8",
                headers={"Content-Disposition": f'inline; filename="{filename}"'},
            )
        if filename.lower().endswith(".json"):
            json_content = scan_result_storage.read_text(filename)
            return Response(
                content=json_content,
                media_type="application/json",
                headers={"Content-Disposition": f'inline; filename="{filename}"'},
            )
        if filename.lower().endswith(".csv"):
            csv_content = scan_result_storage.read_text(filename)
            return Response(
                content=csv_content,
                media_type="text/csv; charset=utf-8",
                headers={"Content-Disposition": f'inline; filename="{filename}"'},
            )
    except ScanResultStorageError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=str(exc),
        ) from exc
    raise HTTPException(
        status_code=status.HTTP_400_BAD_REQUEST,
        detail="Unsupported file type",
    )


@router.post(
    "/markdown",
    response_model=ExportSavedResponse,
    summary="Export quick-scan report as Markdown",
)
def export_markdown(
    request: ExportReportRequest,
    _: Annotated[User, Depends(require_viewer)],
) -> ExportSavedResponse:
    """Generate, save, and return a Markdown analytics report."""
    try:
        report = quick_scan_service.get_report(request.report_id)
    except QuickScanError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=str(exc),
        ) from exc

    markdown = report_service.to_markdown(report)
    saved = scan_result_storage.save_text(
        content=markdown,
        original_filename=report.filename,
        extension="md",
    )
    return ExportSavedResponse(saved=saved, download_filename=saved.filename)


@router.post(
    "/json",
    response_model=ExportSavedResponse,
    summary="Export quick-scan report as JSON",
)
def export_json(
    request: ExportReportRequest,
    _: Annotated[User, Depends(require_viewer)],
) -> ExportSavedResponse:
    """Generate, save, and return a JSON analytics report."""
    try:
        report = quick_scan_service.get_report(request.report_id)
    except QuickScanError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=str(exc),
        ) from exc

    json_report = report_service.to_json(report)
    saved = scan_result_storage.save_text(
        content=json_report,
        original_filename=report.filename,
        extension="json",
    )
    return ExportSavedResponse(saved=saved, download_filename=saved.filename)


@router.post(
    "/csv",
    response_model=ExportSavedResponse,
    summary="Export quick-scan report as CSV",
)
def export_csv(
    request: ExportReportRequest,
    _: Annotated[User, Depends(require_viewer)],
) -> ExportSavedResponse:
    """Generate, save, and return a CSV analytics report."""
    try:
        report = quick_scan_service.get_report(request.report_id)
    except QuickScanError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=str(exc),
        ) from exc

    csv_report = report_service.to_csv(report)
    saved = scan_result_storage.save_text(
        content=csv_report,
        original_filename=report.filename,
        extension="csv",
    )
    return ExportSavedResponse(saved=saved, download_filename=saved.filename)


@router.post(
    "/pdf",
    response_model=ExportSavedResponse,
    summary="Export quick-scan report as PDF",
)
def export_pdf(
    request: ExportReportRequest,
    _: Annotated[User, Depends(require_viewer)],
) -> ExportSavedResponse:
    """Generate, save, and return a PDF analytics report."""
    try:
        report = quick_scan_service.get_report(request.report_id)
    except QuickScanError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=str(exc),
        ) from exc

    pdf_bytes = report_service.to_pdf(report)
    saved = scan_result_storage.save_bytes(
        content=pdf_bytes,
        original_filename=report.filename,
        extension="pdf",
    )
    return ExportSavedResponse(saved=saved, download_filename=saved.filename)
